src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,258 @@
+const std = @import("std");
+const sig = @import("../lib.zig");
+
+const Allocator = std.mem.Allocator;
+const DefaultRwLock = std.Thread.RwLock.DefaultRwLock;
+
+const BytesRef = sig.blockstore.database.BytesRef;
+const Database = sig.blockstore.database.Database;
+const ColumnFamily = sig.blockstore.database.ColumnFamily;
+const Logger = sig.trace.Logger;
+
+pub const SharedHashMapDB = struct {
+    allocator: Allocator,
+    maps: []SharedHashMap,
+    /// shared lock is required to call locking map methods.
+    /// exclusive lock is required to call non-locking map methods.
+    transaction_lock: DefaultRwLock = .{},
+
+    const Self = @This();
+
+    pub const Batch = MapBatch;
+
+    pub fn open(
+        allocator: Allocator,
+        _: Logger,
+        _: []const u8,
+        column_families: []const ColumnFamily,
+    ) !SharedHashMapDB {
+        var maps = try allocator.alloc(SharedHashMap, column_families.len);
+        errdefer allocator.free(maps);
+        errdefer for (maps) |*m| m.deinit();
+        inline for (0..column_families.len) |i| {
+            maps[i] = try SharedHashMap.init(allocator);
+        }
+        return .{ .allocator = allocator, .maps = maps };
+    }
+
+    pub fn deinit(self_: Self) void {
+        var self = self_;
```

Comment:
It might have been for compatibility with Database.deinit. I'm going to mess with that abstraction a bit and see if I can remove the need for weirdness like this.

Reply:
This has been addressed:

https://github.com/Syndica/sig/blob/9be5d821a6487c20131c09f2bc8854d732f5ccfd/src/blockstore/hashmap_db.zig#L200